Dockerfile for creating the EFI image

t-oster opened the pull request:
This is my first attempt to add a Dockerfile and docker-compose.yml
The Dockerfile is not perfect yet and I am not sure if this is the right way with regards to when to build and which directorys to include at docker build time vs docker runtime.

However if you run:
```
git clone -b feature-docker git@github.com:t-oster/zfsbootmenu.git
cd zfsbootmenu
docker-compose up
```
you end up with a fresh vmlinuz.EFI in your ./out directory.

You can then change the sourcecode at your will and rerun
```
sudo rm -rf out #we need to delete the directory because it was created as root
docker-compose up
```
in order to regenerate the image.

ahesford commented on docker/config.yaml:
```diff
@@ -0,0 +1,17 @@
+  Versions: false
+  Enabled: true
+Kernel:
+  CommandLine: ro quiet loglevel=0
```
```yaml
CommandLine: zfsbootmenu ro loglevel=4 nomodeset
```
Use `loglevel=4` to provide some useful warnings if need be, and `nomodeset` to avoid initializing the GPU.

ahesford reviewed (changes requested):
This is an excellent idea; I've wanted to provide some kind of container definition for awhile but haven't been motivated. You've provided a good foundation, but I've made some adjustments that should make this more robust. I'll incorporate my suggestions into a commit that I'll push to your branch.

Note that I expect this to be installed in `contrib/docker`, so the paths have been changed to reflect that. Also, I'll add a `.gitignore` to keep `config.yaml` out of revision control.

Please give the changes a try, make improvements as necessary, and we can squash everything down before merging.

Thanks for your contribution!

t-oster commented on Dockerfile:
```diff
@@ -0,0 +1,13 @@
+#the fisrt cpan install fails because I have no idea how to use it correctly
+#in a dockerfile, but the following ones do work. Somebody please fix.
+RUN cpan install Sort::Versions || cpan install Sort::Versions Config::IniFiles YAML::PP boolean
+ADD . /zbm
```
As already said I would not git pull it again because I suppose the workflow is more like check out the source, maybe modify it and then use docker to build it, so the source is already checked out and should be used. However I think the during build you can only use the contents of the current directory, so the Dockerfile must be in the root directory or we need the context directive in docker-compose https://stackoverflow.com/questions/24537340/docker-adding-a-file-from-a-parent-directory

ahesford commented:
Thinking about this a bit more, there is a more generic way to approach the build. I think it would be good to make the built image as reusable as possible, so let's not copy the local clone or create a new clone of the upstream repo at build time. Instead, we'll just make a basic Void image with all of the pieces necessary to do the build.

The build can be handled by a `zbm-build.sh` script which will check `/zbm` to make sure it looks like it has the components we need for a build. If so, it will happily run the build as normal. If `/zbm` is empty, it will instead install `git` and then clone the upstream repo. This means the image is totally usable without requiring a local repo, but you can bind-mount the local repo to /zbm and that will be used instead.

What do you think about the script in the latest commit?

t-oster commented:
Hi, this looks good to me, because I can still build the current directory with just docker-compose up and we don't add/clone the source twice. Very elegant.
If you want you can squash and merge it or do you need anything from me?

t-oster commented:
Hi, I squashed the commits, but so your messages/name vanished, so I added a Co-Authored line. However if you prefer three commits I can also revert that, whatever you like better.

ahesford commented:
I made a few more fixes and pushed 416696773a91429e51e6d0db4fcdb23f37b89a8c as the final version. Specifically:
- A README provides usage information
- The build script now looks for `hostid` and `zpool.cache` in the `contrib/docker` directory and will populate these files in the container so they are installed in the ZBM images

Thanks for your contributions. This will be very helpful to our release process!

t-oster commented:
You're welcome. thanks for reviewing and merging so quickly and again for this great tool, which makes me feel safe again when running zpool upgrade on my server.
